app/backend/src/couchers/servicers/search.py

```diff
@@ -514,3 +523,91 @@ def UserSearch(self, request, context):
+                            EventOccurrence.address.ilike(f"%{request.query.value}%"),
+                        )
+                    )
+
+            if not request.include_cancelled:
+                statement = statement.where(~EventOccurrence.is_cancelled)
+
+            if request.HasField("search_in_area"):
+                # EPSG4326 measures distance in decimal degress
+                # we want to check whether two circles overlap, so check if the distance between their centers is less
+                # than the sum of their radii, divided by 111111 m ~= 1 degree (at the equator)
+                search_point = create_coordinate(request.search_in_area.lat, request.search_in_area.lng)
+                statement = statement.where(
+                    func.ST_DWithin(
+                        # old:
+                        # User.geom, search_point, (User.geom_radius + request.search_in_area.radius) / 111111
+                        # this is an optimization that speeds up the db queries since it doesn't need to look up the user's geom radius
+                        EventOccurrence.geom,
+                        search_point,
+                        (1000 + request.search_in_area.radius) / 111111,
+                    )
+                )
+            if request.HasField("search_in_rectangle"):
+                statement = statement.where(
+                    func.ST_Within(
+                        EventOccurrence.geom,
+                        func.ST_MakeEnvelope(
+                            request.search_in_rectangle.lng_min,
+                            request.search_in_rectangle.lat_min,
+                            request.search_in_rectangle.lng_max,
+                            request.search_in_rectangle.lat_max,
+                            4326,
+                        ),
+                    )
+                )
+            if request.HasField("search_in_community_id"):
+                # could do a join here as well, but this is just simpler
+                node = session.execute(
+                    select(Node).where(Node.id == request.search_in_community_id)
+                ).scalar_one_or_none()
+                if not node:
+                    context.abort(grpc.StatusCode.NOT_FOUND, errors.COMMUNITY_NOT_FOUND)
+                statement = statement.where(func.ST_Contains(node.geom, EventOccurrence.geom))
+
+            if request.HasField("after"):
+                statement = statement.where(EventOccurrence.start_time > to_aware_datetime(request.after))
+            if request.HasField("before"):
+                statement = statement.where(EventOccurrence.end_time < to_aware_datetime(request.before))
+
+            page_size = min(MAX_PAGINATION_LENGTH, request.page_size or MAX_PAGINATION_LENGTH)
+            # the page token is a unix timestamp of where we left off
+            page_token = dt_from_millis(int(request.page_token)) if request.page_token else now()
```

Comment:
Sorry to be annoying but could you add a test to double check pagination works correctly both forwards and backwards please?

Reply:
Added tests for it now 👍